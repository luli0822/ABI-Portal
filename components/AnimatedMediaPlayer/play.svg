<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="40" height="40" viewBox="0 0 40 40">
    <defs>
        <path id="a" d="M20 0C8.973 0 0 8.973 0 20s8.973 20 20 20 20-8.973 20-20S31.027 0 20 0zm0 36.436c-9.057 0-16.436-7.38-16.436-16.436 0-9.057 7.38-16.436 16.436-16.436 9.057 0 16.436 7.38 16.436 16.436 0 9.057-7.38 16.436-16.436 16.436zm-3.648-24.36c-.461-.252-1.048.083-1.048.586l-.042 14.634c0 .545.587.88 1.048.629l12.705-7.296c.46-.252.46-.923 0-1.216l-12.663-7.338z"/>
    </defs>
    <g fill="none" fill-rule="evenodd">
        <mask id="b" fill="#fff">
            <use xlink:href="#a"/>
        </mask>
        <use fill="#002F6C" fill-rule="nonzero" xlink:href="#a"/>
        <g fill="#FFF" mask="url(#b)">
            <path d="M-47-48H97V96H-47z"/>
        </g>
    </g>
</svg>
