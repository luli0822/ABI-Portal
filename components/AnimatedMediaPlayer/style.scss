$root-class: "animated-media-player";

.#{$root-class} {
  position: relative;

  &__video {
    display: block;
    width: 100%;
  }

  &__overlay {
    display: flex;
    position: absolute;
    width: 100%;
    height: 100%;
    align-items: center;
    justify-content: center;
    background-color: rgba(255, 255, 255, 0.253);

    &--hidden {
      display: none;
    }
  }
}
