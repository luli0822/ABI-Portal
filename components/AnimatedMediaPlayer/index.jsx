import React, { useRef, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import PlayButton from './PlayButton';
import './style.scss';

function AnimatedMediaPlayer(props) {
	const { mp4Url, posterUrl, autoplay, description } = props;
	const rootClass = 'animated-media-player';
	const videoRef = useRef(null);

	// values for isPlaying include 'autoplay', 'hover', 'click', and 'keyDown'
	// to determine what started the video
	const [isPlaying, setIsPlaying] = useState(autoplay ? 'autoplay' : false);

	// controls whether the video should play again when the first autoplay ends
	const [autoPlaysRemaining, setAutoPlaysRemaining] = useState(autoplay - 1);

	useEffect(() => {
		if (autoplay) {
			const videoElement = videoRef.current;
			// don't use the play() function because that sets loop = true, but autoplay shouldn't loop
			videoElement.play();
			// we don't need to set isPlaying because it's already set to 'autoplay' above
		}
	}, [autoplay]);

	/**
	 * Start playing the video. If loop==true, set the video to continually loop
	 * @param {string} trigger
	 * @param {boolean} loop
	 */
	function play(trigger, loop) {
		const videoElement = videoRef.current;
		if (videoElement.paused) {
			// only start the video and set the trigger if it's not already playing
			videoElement.play();
			videoElement.loop = loop;
			setIsPlaying(trigger);
		}
	}

	function pause() {
		const videoElement = videoRef.current;
		videoElement.pause();
		setIsPlaying(false);
	}

	function togglePlay(trigger) {
		const videoElement = videoRef.current;
		if (videoElement.paused) {
			play(trigger, true);
		} else {
			pause();
		}
	}

	function clickHandler() {
		togglePlay('click');
	}

	function keyDownHandler(event) {
		if (event.key === 'Enter') {
			togglePlay('keyDown');
		}
	}

	function mouseOverHandler() {
		play('hover', true);
	}

	function mouseOutHandler() {
		// only pause the video if the trigger was mouse-over
		if (isPlaying === 'hover') {
			pause();
		}
	}

	function endedHandler() {
		setIsPlaying(false);
		if (autoPlaysRemaining > 0) {
			setAutoPlaysRemaining(autoPlaysRemaining - 1);
			play('autoplay');
		}
	}

	return (
		<div className={rootClass}>
			<div className={`${rootClass}__overlay ${isPlaying ? `${rootClass}__overlay--hidden` : ''}`}>
				<PlayButton />
			</div>
			{/* eslint-disable-next-line jsx-a11y/mouse-events-have-key-events */}
			<video
				muted
				poster={posterUrl}
				className={`${rootClass}__video`}
				tabIndex={0}
				ref={videoRef}
				onClick={clickHandler}
				onKeyDown={keyDownHandler}
				onMouseOver={mouseOverHandler}
				onMouseOut={mouseOutHandler}
				onEnded={endedHandler}
				aria-label={description}
			>
				<source src={mp4Url} type="video/mp4" />
			</video>
		</div>
	);
}

AnimatedMediaPlayer.propTypes = {
	mp4Url: PropTypes.string.isRequired,
	posterUrl: PropTypes.string.isRequired,
	description: PropTypes.string.isRequired,
	autoplay: PropTypes.number,
};

AnimatedMediaPlayer.defaultProps = {
	autoplay: 0,
};

export default AnimatedMediaPlayer;
