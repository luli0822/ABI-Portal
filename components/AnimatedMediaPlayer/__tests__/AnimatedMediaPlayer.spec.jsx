import React from 'react';
import { shallow } from 'enzyme';
import AnimatedMediaPlayer from '../index';

describe('<AnimatedMediaPlayer />', () => {
	it('renders correctly', () => {
		// https://github.com/facebook/jest/issues/7802
		// using .debug() instead of adding dependency on enzyme-to-json for now
		expect(
			shallow(
				<AnimatedMediaPlayer mp4Url="video.mp4" posterUrl="poster.jpg" description="A video" />
			).debug()
		).toMatchSnapshot();
	});
});
