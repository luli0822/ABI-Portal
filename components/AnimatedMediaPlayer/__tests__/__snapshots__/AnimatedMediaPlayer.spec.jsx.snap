// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnimatedMediaPlayer /> renders correctly 1`] = `
"<div className=\\"animated-media-player\\">
  <div className=\\"animated-media-player__overlay \\">
    <PlayButton />
  </div>
  <video muted={true} poster=\\"poster.jpg\\" className=\\"animated-media-player__video\\" tabIndex={0} onClick={[Function: clickHandler]} onKeyDown={[Function: keyDownHandler]} onMouseOver={[Function: mouseOverHandler]} onMouseOut={[Function: mouseOutHandler]} onEnded={[Function: endedHandler]} aria-label=\\"A video\\">
    <source src=\\"video.mp4\\" type=\\"video/mp4\\" />
  </video>
</div>"
`;
